File: wwwroot/js/reservation.js
// Please see documentation at https://docs.microsoft.com/aspnet/core/client-side/bundling-and-minification
// for details on configuring this project to bundle and minify static web assets.

let checkInField = null;
let checkOutField = null;
let priceField = null;


document.addEventListener("DOMContentLoaded", (event) => {
    priceField = document.getElementById("totalPrice");
    checkInField = document.getElementById("checkIn");
    checkOutField = document.getElementById("checkOut");

    checkInField.addEventListener("change", validateDates);
    checkOutField.addEventListener("change", validateDates);

    updatePrice();

});

function updatePrice() {
    if (checkInField != null && checkOutField != null) {
        let checkInDate = new Date(checkInField.value);
        let checkOutDate = new Date(checkOutField.value);
        let difference = checkOutDate.getTime() - checkInDate.getTime();
        let totalDays = Math.ceil(difference / (1000 * 3600 * 24));

        let roomSelect = document.getElementById("roomSelect");
        let roomOption = roomSelect.options[roomSelect.selectedIndex].text;

        let dailyPrice = roomOption.slice(1, roomOption.search("/"));

        let totalPrice = totalDays * dailyPrice;
        if (!isNaN(totalPrice)) {
            priceField.value = "$" + totalPrice;

            let hiddenPriceField = document.getElementById("Price");
            hiddenPriceField.value = totalPrice;
        }

    }
}

function validateDates() {
    if (checkInField != null && checkOutField != null) {
        let currentDate = new Date();

        let checkInDate = new Date(checkInField.value);
        let checkOutDate = new Date(checkOutField.value);

        let dateDifference = (checkOutDate.getTime() - checkInDate.getTime()) / 1000 / 3600;

        let infoBar = document.getElementById("infoBar");
        if (checkInDate >= checkOutDate) {
            infoBar.innerHTML = "Check-in date must be before check-out date.";
        }
        else if (checkInDate < currentDate.getTime() || checkOutDate < currentDate.getTime()) {
            infoBar.innerHTML = "Check-in and check-out dates must be in the future.";
        }
        else if (dateDifference < 8) {
            infoBar.innerHTML = "Reservation must be at least 8 hours long.";
        }
        else {
            infoBar.innerHTML = "";
        }
    }
}

File: Models/Entities/Reservation.cs
using WebAppHotel.Models.Enums;

namespace WebAppHotel.Models.Entities
{
    public class Reservation
    {
        public int Id { get; set; }
        public int RoomId { get; set; }
        public int CustomerId { get; set; }
        public DateTime CheckIn { get; set; }
        public DateTime CheckOut { get; set; }
        public double Price { get; set; }
        public ReservationStatus Status { get; set; }

        public Reservation(int id, int customerId, int roomId, DateTime checkIn, DateTime checkOut, double price, ReservationStatus status)
        {
            if (IsDateValid(checkIn, checkOut))
            {   
                Id = id;
                CustomerId = customerId;
                RoomId = roomId;
                CheckIn = checkIn;
                CheckOut = checkOut;
                Price = price;
                Status = status;
            }
        }

        public Reservation() { }

        public static bool IsDateValid(DateTime checkIn, DateTime checkOut)
        {
            if (checkIn >= checkOut) 
            {
                throw new ArgumentException("Check-in date must be before check-out date.");
            }
            if (checkIn < DateTime.Now || checkOut < DateTime.Now)
            {
                throw new ArgumentException("Check-in and check-out dates must be in the future.");
            }
            if (checkOut.Subtract(checkIn).Hours < 8)
            {
                throw new ArgumentException("Reservation must be at least 8 hours long.");
            }
            return true;
        }
    }
}
